Add UserGuide documentation for `IMDReader` (#430)
MDAnalysis/UserGuide: doc/source/formats/reference/imd.rst

Reviewer: General comment: link to the primary sources (imdclient docs, IMDReader docs)

.. -*- coding: utf-8 -*-
.. _IMD-format:

==================================================================
IMD (Data streamed via Interactive Molecular Dynamics protocol v3)
==================================================================

.. include:: classes/IMD.txt

> Real-time streaming of simulation data between molecular dynamics engines and receiving clients can be achieved using IMD protocols like IMDv2 and IMDv3. The :class:`~MDAnalysis.coordinates.IMD.IMDReader` implements the IMDv3 protocol, enabling live streaming of ongoing simulation data.

Reviewer 2: ```suggestion
IMDv2 and IMDv3 enable real-time streaming of simulation data between molecular dynamics engines and receiving clients. The :class:`~MDAnalysis.coordinates.IMD.IMDReader` implements the IMDv3 protocol.
```

Reviewer: @ljwoods2 suggestion is an improvement (so I am adding it) but it's still too much jargon for an opening paragraph — see my comments above.

> Real-time streaming of simulation data between molecular dynamics engines and receiving clients can be achieved using IMD protocols like IMDv2 and IMDv3. The :class:`~MDAnalysis.coordinates.IMD.IMDReader` implements the IMDv3 protocol, enabling live streaming of ongoing simulation data.

Reviewer: You need to explain what the general purpose is (re-summarize paper, IMDReader, imdclient docs). Put yourself in the position of someone whose **first contact** with the idea is through the User Guide.  Answer the questions

* What is it?
* Why is it useful?
* Where can I learn more?

Also explain abbreviation "IMD". 

Recyle what you've already have written, e.g. from the "What is streaming" section, which is already great in answering the first question. You just need to taylor the initial 2-3 sentences so that they make sense on their own.

Author: Done!

.. note::

>    MDAnalysis supports **IMDv3 only**, which provides continuous, gap-free streaming and is implemented in modern versions of GROMACS, LAMMPS, and NAMD. IMDv2, while widely available, was designed primarily for visualization and allows gaps in the data stream.

Reviewer 2: ```suggestion
   MDAnalysis supports **IMDv3 only**, which provides continuous, gap-free streaming and is implemented in modern versions of GROMACS, LAMMPS, and NAMD. IMDv2, while widely available, was designed primarily for visualization and doesn't enforce a consistent number of integration steps between transmitted frames
```

What is Streaming?
==================

> Streaming involves processing data in real-time as it is generated, rather than storing it for later analysis. In molecular dynamics, this means sending simulation data to a client on-the-fly while the simulation is running, without writing large trajectory files to disk.

Reviewer: This is good!

> This can be achieved through a TCP/IP socket connection between the simulation engine and receiving client, transmitting coordinates, velocities, forces, energies, and timing information using the IMDv3 protocol.

Reviewer 2: ```suggestion
In IMDv3, this is achieved through a TCP/IP socket connection between the simulation engine and receiving client, transmitting coordinates, velocities, forces, energies, and timing information.
```

MDAnalysis's :class:`~MDAnalysis.coordinates.IMD.IMDReader` uses the `imdclient <https://imdclient.readthedocs.io/>`_ package and provides a familiar interface for reading streaming data, similar to other trajectory readers in MDAnalysis.

> When to Use Streaming?

Reviewer: very nice — answers my questions from the first paragraph: so summarize in the first paragraph at a high level what you're writing here.

======================

Streaming analysis is particularly valuable for:

**Long-running simulations**
    Early detection of problems (crashes, artifacts, equilibration issues) can save computational resources.

**Adaptive sampling workflows**
    Real-time analysis can guide simulation parameters or trigger enhanced sampling methods.

**Interactive research**
    Immediate feedback allows researchers to make informed decisions about continuing, modifying, or terminating simulations.

**Storage-constrained environments**
    Analyze data as it's generated without storing large trajectory files.

> Installation and Setup

Reviewer: Explain that imdclient is an _optional_ package and has to be installed separately from MDAnalysis, either with mamba or pip.

I'd just condense the whole section and link to the authoritative https://imdclient.readthedocs.io/en/stable/getting_started.html#installation

Author: Done!

======================

Required Dependencies
---------------------

> The IMDReader requires the ``imdclient`` package:

Reviewer: Link imdclient to its docs.

Author: Done!

.. code-block:: bash

>    pip install imdclient

Reviewer 2: I think we encourage conda/mamba installation

Reviewer: yes!

Author: Made the required change!

.. note::
   MDAnalysis requires ``imdclient >= 0.2.2`` for its current implementation.

> MD Engine Configuration

Reviewer: Definitely link to the imdclient docs that say more or less the same thing.

Reviewer: I'd also suggest to _not_ repeat this configuration in the notebook and just link from the notebook to here.

Author: Done!

-----------------------

> We provide below example configurations for enabling IMDv3 streaming in popular MD engines.

Reviewer: To make this section different from the authoritative https://imdclient.readthedocs.io/en/stable/usage.html#configuring-the-simulation-engine-for-imdv3 I suggest that you give an _actual example_ configuration for each engine with real example values and not place holders like `<yes/no>`.

Imagine that students (and LLMS!) will use this information in order to get started.

If you were a PI, what values would you want your student to use? Put these in and add a sentence or two explaining your reasoning.

> **GROMACS**

Reviewer: Make this a heading of the right level.

It's bad practice to use emphasis markup instead of semantic markup for structuring documents. Not only does it make formatting inconsistent (using styles) but it also makes documents hard to read on screen readers and thus decreases the accessibility dramatically.

Author: Done!

> Add IMD settings to your ``.mdp`` file:

Reviewer: Add text that _explains_ what these settings do, or at minimum highlight the important settings.

Definitely provide a link to the imdclient docs (which should be the authoritative source).

Author: Shouldn't the MD engine docs be the source here?

.. code-block:: text

    ; IMD settings
    IMD-group        = System
    IMD-version      = 3

>     IMD-nst          = 1

Reviewer: perhaps not every step? what would you choose?

Author: maybe 100?

>     IMD-time         = No

Reviewer: ```suggestion
    IMD-time         = Yes
```

Author: Done!

    IMD-coords       = Yes

>     IMD-vels         = No

Reviewer: In the example also stream velocities.

Author: Done!

    IMD-forces       = No

>     IMD-box          = No

Reviewer: Being able to do this was almost the whole point of IMDv3
```suggestion
    IMD-box          = Yes
```
(+ velocities)

Author: Done!

    IMD-unwrap       = No

>     IMD-energies     = No

Reviewer: and transmit energies

Author: So tramsit everything in the example, then?

Run with IMD enabled:

.. code-block:: bash

    gmx mdrun -v -nt 4 -imdwait -imdport 8889

> **LAMMPS**

Reviewer: heading

Author: Done!

Use the IMD fix in your input script:

> .. code-block:: text

Reviewer: Make the example equivalent to the GROMACS one.

Author: Done!

    # IMD setup
    fix ID group-ID imd <port> trate <frequency> version 3 unwrap <on/off> time <on/off> box <on/off> coordinates <on/off> velocities <on/off> forces <on/off>

Run your LAMMPS simulation as usual.

> **NAMD**

Reviewer: heading

Author: Done!

Add IMD configuration to your NAMD configuration file:

> .. code-block:: text

Reviewer: Make the example equivalent to the GROMACS one.

Author: Done!

    # IMD Settings
    IMDon yes
    IMDport <port, must be the same port used for the client>
    IMDwait <yes/no>
    IMDfreq <frequency of sending data to the client>

    IMDsendPositions <yes/no>
    IMDsendEnergies <yes/no>
    IMDsendTime <yes/no>
    IMDsendBoxDimensions <yes/no>
    IMDsendVelocities <yes/no>
    IMDsendForces <yes/no>
    IMDwrapPositions <yes/no>

Run your NAMD simulation as usual.

.. seealso::
   For detailed engine-specific setup instructions, see the `imdclient simulation engine documentation <https://imdclient.readthedocs.io/en/latest/usage.html>`_.

Basic Usage
===========

Connecting to a Running Simulation
-----------------------------------

> Once your simulation is running with IMD enabled:

Reviewer: explain choice of buffer_size.

Author: Done!

.. code-block:: python

   import MDAnalysis as mda
   
   # Connect to the simulation
   u = mda.Universe("topol.tpr", "imd://localhost:8889", buffer_size=10*1024*1024)
   
   # Streaming analysis loop
   for ts in u.trajectory:
       print(f"Time: {ts.time:.2f} ps, Step: {ts.data.get('step', 'N/A')}")
       
       # Your analysis code here

>        selected_atoms = u.select_atoms("protein and name CA")

Reviewer 3: to clarify -- is there something about IMD that mandates repeating the selection of the `AtomGroup` in the loop, or can this be hoisted up for efficiency as usual?

Reviewer 2: Nothing unique about IMD here, can be changed

Reviewer: @amruthesht move selection out of the loop to model best practices (it's bad for performance!)

Author: @tylerjereddy - nothing specific about IMD requires it. I'll modify the example so it doesn't give the wrong impression

       center_of_mass = selected_atoms.center_of_mass()
       print(f"Protein COM: {center_of_mass}")
       
       # Optional: break on some condition

>        if ts.time > 1000:  # Stop after 1000 ps

Reviewer 3: to clarify: this *only* stops the analysis, not the running simulation/engine the analysis is connected to?

Reviewer 2: Since this break is the end of the script, it will trigger the IMDClient to cleanup via atext

https://github.com/Becksteinlab/imdclient/blob/07069e1ff08d1488936c61e52f922116ecddf210/imdclient/IMDClient.py#L146

That cleanup might put the simulation engine in a waiting state for another client connection, but won't kill it.

Whether it makes the simulation engine wait or not depends on 1. how the simulation engine was configured and 2. whether the IMDClient overrode the simulation engine's waiting behavior

`continue_after_disconnect` kwarg is the IMDClient switch for changing waiting behavior

https://imdclient.readthedocs.io/en/latest/api.html

Reviewer 3: I wonder if the need to look out for this/think about this might be sensible to mention somewhere inline here then, if only to refer to more detailed docs for someone who may be copy/pasting the sample code as a template.

Reviewer: I'd remove the optional part from this example so that it can stand as a simple example on its own.

Then create another example where you discuss the issue with how breaking the loop may affect the running simulation. This is important but nuanced. If something is complicated, break it into smaller steps, that's basic pedagogy.

Reviewer: Would still recommend breaking up this example

> I'd remove the optional part from this example so that it can stand as a simple example on its own.
> 
> Then create another example where you discuss the issue with how breaking the loop may affect the running simulation. This is important but nuanced. If something is complicated, break it into smaller steps, that's basic pedagogy.

           break

Real-time Quality Control
-------------------------

Monitor simulation health in real-time:

.. code-block:: python

   import MDAnalysis as mda
   import numpy as np
   
   u = mda.Universe("system.tpr", "imd://localhost:8889")
   
   previous_positions = None
   
   for ts in u.trajectory:
       current_positions = u.atoms.positions.copy()
       
       # Check for simulation artifacts
       if previous_positions is not None:

>            displacement = np.linalg.norm(current_positions - previous_positions, axis=1)

Reviewer: The simple distance calculation fails on PBC crossing.

Use PBC-aware distance calculations with [`lib.distances.calc_bonds()`](https://docs.mdanalysis.org/stable/documentation_pages/lib/distances.html#MDAnalysis.lib.distances.calc_bonds). This will also demonstrate why it's important that we transfer box information!

Author: Done!

           max_displacement = np.max(displacement)
           
           if max_displacement > 10.0:  # Atoms moved > 10 Å in one step
               print(f"WARNING: Large displacement detected at {ts.time} ps: {max_displacement:.2f} Å")
       
       # Monitor energies if available

>        if 'potential' in ts.data:

Reviewer 2: I think the key is "potential_energy" 

https://github.com/Becksteinlab/imdclient/blob/07069e1ff08d1488936c61e52f922116ecddf210/imdclient/IMDClient.py#L956

Reviewer: For these examples I might not even build in safe guards but just state at the top that you're assuming that energy is transmitted.

Author: Done!

           print(f"Potential energy: {ts.data['potential']:.2f}")
       
       previous_positions = current_positions

Advanced Features
=================

Buffer Management
-----------------

> For compute-intensive analysis, increase the buffer size to reduce communication overhead:

Reviewer: Link to imdclient docs that (hopefully) have more explanations about what the buffer does.

Author: Done!

.. code-block:: python

   # Larger buffer for better performance

>    u = mda.Universe("topol.tpr", "imd://localhost:8889", buffer_size=50*1024*1024)

Reviewer 2: Specify unit, these are bytes

Author: Added as a comment!

> Connection Management

Reviewer: This is a good section.

If you have other try/except or specialized handling then you can collect it here.

---------------------

Always ensure proper cleanup, especially in interactive environments like Jupyter notebooks et al.:

.. code-block:: python

   import MDAnalysis as mda
   
   try:
       u = mda.Universe("topol.tpr", "imd://localhost:8889")
       
       for ts in u.trajectory:
           # Your analysis here
           pass
           

>    except Exception as e:

Reviewer: Catching all exceptions has the potential to make errors go un-noticed.

How about re-raising the exception after closing the trajecorty? Along the lines of
```python
try:
   ...
except Exception as err:
   # log error
   pass
finally:
   u.trajectory.close()

if err:    # won't work, but that's the logic
   raise err
```

Author: Done!

       print(f"Error during streaming: {e}")
   finally:
       # Always close the connection
       u.trajectory.close()

Available Data
--------------

The IMDReader provides access to additional simulation data through ``ts.data``:

* ``dt``: Time step size in picoseconds
* ``step``: Current simulation step number  

> * Energy terms: ``potential``, ``total``, etc. (engine-dependent)

Reviewer 2: I think these keys are wrong, see:

https://github.com/Becksteinlab/imdclient/blob/07069e1ff08d1488936c61e52f922116ecddf210/imdclient/IMDClient.py#L956

I think we actually merged them into MDA develop docstring for IMDReader incorrectly:

https://imdclient.readthedocs.io/en/latest/protocol_v3.html#energies

Reviewer 2: Also, are these engine-dependent? It's engine depended whether the energy packet is implemented, but the exact energy keys should be consistent whenever the packet is implemented, right?

Reviewer: @ljwoods2 can you raise an issue, please?

Author: @ljwoods2 - yes, the keys should be the same for all engines, only implementation availability is engine dependent - I chnage the langauage around it

Thanks for pointing out the erros in the API documentation. I'll make a PR to correct it!

.. code-block:: python

   for ts in u.trajectory:
       print(f"Step {ts.data.get('step')}: dt={ts.data.get('dt')} ps")
       
       # Available energy terms vary by MD engine
       for key, value in ts.data.items():
           if key not in ['dt', 'step']:
               print(f"  {key}: {value}")

> Multiple Client Connections

Reviewer 2: There's a ton of edge cases when using multiple clients that we don't test, like when one client tries to change blocking behavior, so I think we should just say it isn't recommended

Discussion here

https://github.com/Becksteinlab/imdclient/pull/114 

Maybe should update MDA main codebase docs to say this as well:

https://github.com/amruthesht/mdanalysis/blob/ea85bb510f684ef1f6048ad0a50219665298cbb3/package/MDAnalysis/coordinates/IMD.py#L108

Reviewer: I would leave out the whole section; perhaps mention somewhere that it may be possible but results are not well defined and link to imdclient docs.

Author: @jwoods2 - I actually tested some edge cases out - I'll add it to the relevant PR discussion.

@orbeckst - Done!

===========================

The ability to connect multiple clients to the same IMD port depends on the MD engine implementation:

* **GROMACS**: Typically supports single client connections
* **LAMMPS**: May support multiple clients (version-dependent)
* **NAMD**: Supports multiple clients

.. important::
   Even when multiple connections are supported, each receives an independent data stream. Different clients may receive different data depending on the engine configuration.

Integration with MDAnalysis Tools
=================================

Most MDAnalysis analysis classes work with streaming data, but some limitations apply:

> **Compatible Analysis**

Reviewer 2: I think this should start by outlining what kinds of analyses will and won't work before jumping into distance and contact analysis

Author: Done - could you check if this is accurate, thanks!

> **Compatible Analysis**

Reviewer: use proper section markup for subheadings

Author: Done!

.. code-block:: python

   from MDAnalysis.analysis import distances, contacts
   
   u = mda.Universe("system.tpr", "imd://localhost:8889")
   
   for ts in u.trajectory:
       # Distance calculations work normally
       protein = u.select_atoms("protein")
       rg = protein.radius_of_gyration()
       
       # Contact analysis
       selection1 = u.select_atoms("resid 1-10")

>        selection2 = u.select_atoms("resid 50-60")

Reviewer 3: hoist selections if IMD allows it?

Author: Done!

       dist_array = distances.distance_array(selection1.positions, selection2.positions)

> **Limitations with Streaming**

Reviewer: use proper section markup for subheadings

Author: Done!

Some analysis methods require the complete trajectory and won't work with streaming:

.. code-block:: python

   # These will NOT work with streaming:
   # - trajectory.timeseries()
   # - Most analysis classes that need multiple passes
   # - Random frame access (trajectory[10])

>    # - Backward iteration

Reviewer: Should be a simple list, not a code block

Author: Fixed!

Important Limitations
=====================

Streaming analysis has fundamental constraints due to its real-time nature:

> **Data Access Limitations**

Reviewer: use proper section markup for subheadings

Author: Done!

* **No random access**: Cannot jump to arbitrary frames or seek backwards
* **Forward-only**: Can only iterate through frames as they arrive
* **Single-use**: Cannot restart iteration once the stream is consumed
* **No trajectory length**: Total frame count unknown until simulation ends
* **No independent copies**: Cannot create multiple reader instances for the same stream

> **Analysis Constraints**

Reviewer: use proper section markup for subheadings

Author: Done!

* **No timeseries methods**: Cannot use ``trajectory.timeseries()``
* **No bulk operations**: Cannot extract all data at once

> * **Limited multiprocessing**: Cannot split across processes

Reviewer 3: Is there a concrete example of something you can do with `multiprocessing` with "vanilla" MDAnalysis that doesn't work with IMD? I would have figured that if you can get data into a NumPy array you're probably "ok" to split off additional processes if you want to.

Reviewer 2: By vanilla MDA, does that exclude analysis classes with dask/multiprocessing backend?

You could copy the entire traj into np array as it goes, then create universe with traj from array, but that would defeat the purpose I think 

And you can use multiprocessing/parallelism on a single frame / sliding window of frames, I know @HeydenLabASU has tried some windowing and then using parallel algorithms to process an i.e. 200 frame window

I guess maybe this should be changed to say that it only works with the serial analysis backend

Reviewer 3: >  that would defeat the purpose I think

I guess you could "buffer"/store a few frames in an array that way without doing IO, and possibly only for a select subset of particles, which does seem within the remit/spirit of what IMD enables, or possibly even one of the more interesting use cases. I think that's what you're saying Matthias is doing.

I might guess that accessing multiple frames in parallel is a problem for the same reason that random access is not available, and that some analyses that require multiple passes and divide work over random-access frame access are also banned, so yeah just trying to think of cases that block multiprocessing usage that aren't already covered by the elementary limitations mentioned. Maybe the multiple clients in different processes thing, but I think you also already ruled them out in comments above, whether in different processes or not.

* **Single client**: Only one reader per IMD stream (engine-dependent)

> **Practical Considerations**

Reviewer: use proper section markup for subheadings

Author: Done!

.. code-block:: python

   # This WILL work - forward iteration
   for ts in u.trajectory:
       analysis_data.append(calculate_something(ts))
   
   # This will NOT work - random access

>    ts = u.trajectory[10]  # TypeError

Reviewer 2: From reading docstring of `__getitem__` of StreamReaderBase, I think these should both be ValueErrors

``` 
def __getitem__(self, frame):
        """Return an iterator for slicing a streaming trajectory.

        Parameters
        ----------
        frame : slice
            Slice object. Only the step parameter is meaningful for streams.

        Returns
        -------
        FrameIteratorAll or StreamFrameIteratorSliced
            Iterator for the requested slice.

        Raises
        ------
        TypeError
            If frame is not a slice object.
        ValueError
            If slice contains start or stop values.
```

If it does throw a TypeError, something is wrong...

Author: Agreed, thanks for pointing it out!

>    ts = u.trajectory[10]  # TypeError

Reviewer: split into separate code blocks with preceding text that states what is being attempted and what will happen; the comments "will work/won't work" can also stay to make it clear.

Author: Done!

   
   # This will NOT work - backwards iteration  
   for ts in u.trajectory[::-1]:  # ValueError
       pass
   
   # This will NOT work - restarting iteration
   for ts in u.trajectory:
       break

>    for ts in u.trajectory:  # Won't start from beginning

Reviewer 2: This suggest it will start from somewhere else other than the beginning, won't it just fail?

Author: Wouldn't it just start from the next available frame in the buffer?

Reviewer 2: No, it shouldn't if I'm understanding the code right (just trying to run a for loop two times), it should raise an error:

https://github.com/MDAnalysis/mdanalysis/blob/03eef459ddf83e73a776c0e6ddefde4a04ce62c9/testsuite/MDAnalysisTests/coordinates/test_imd.py#L457

>        pass

Reviewer: ALso add the `u.trajectory[:10]` example: we cannot (currently) set end-frames.

Author: Done!

    IMDsendPositions yes    ; Send coordinates (yes/no)
    IMDsendVelocities yes   ; Send velocity data (yes/no)
    IMDsendForces yes       ; Send force data (yes/no)
    
    # Coordinate processing
    IMDwrapPositions no    ; Don't wrap (i.e unwrap) coordinates into simulation box (yes/no)

Run your NAMD simulation as usual.

Basic Usage
===========

Connecting to a Running Simulation
-----------------------------------

Once your simulation is running with IMD enabled:

.. code-block:: python

   import MDAnalysis as mda
   
   # Connect to the simulation
   u = mda.Universe("topol.tpr", "imd://localhost:8889", buffer_size=10*1024*1024)

   # Select atoms for analysis
   selected_atoms = u.select_atoms("protein and name CA")
   
   # Streaming analysis loop
   for ts in u.trajectory:
       print(f"Time: {ts.time:.2f} ps, Step: {ts.data.get('step', 'N/A')}")
       
       # Your analysis code here
       center_of_mass = selected_atoms.center_of_mass()
       print(f"Protein COM: {center_of_mass}")
       
       # Optional: break on some condition
       if ts.time > 1000:  # Stop after 1000 ps
           break

> The ``buffer_size`` parameter (10 MB = 10*1024*1024 bytes in this example) controls the buffer used by `imdclient <https://imdclient.readthedocs.io/>`_ to temporarily store data received from the socket. This buffer accounts for speed differences between the producer (simulation engine) and receiver (analysis code), preventing data loss when analysis is slower than data transmission. A larger buffer is more suitable for systems with many atoms or high transmission frequencies. For more details on buffer management and optimization, see the `imdclient documentation <https://imdclient.readthedocs.io/>`_.

Reviewer: Generally I would try to replace in-line hyperlinks with named links just to keep things tidy, especially if you're using the same URL multiple times. (Optional)

   u = mda.Universe("topol.tpr", "imd://localhost:8889", buffer_size=50*1024*1024)

For detailed information about buffer behavior and usage, see the `imdclient buffer management documentation <https://imdclient.readthedocs.io/>`_.

Connection Management
---------------------

Always ensure proper cleanup, especially in interactive environments like Jupyter notebooks and other interactive environments:

.. code-block:: python

   import MDAnalysis as mda
   
   u = None
   error = None
   
   try:
       u = mda.Universe("topol.tpr", "imd://localhost:8889")
       
       for ts in u.trajectory:
           # Your analysis here
           pass
           
   except Exception as e:
       # Log error but don't re-raise yet
       error = e
       print(f"Error during streaming: {e}")

   finally:
       # Always close the connection first
       if u is not None:
           u.trajectory.close()
   
   # Re-raise after cleanup is done
   if error:
       raise error

Available Data
--------------

The IMDReader provides access to additional simulation data through ``ts.data``:

* ``dt``: Time step size in picoseconds
* ``step``: Current simulation step number

> * Energy terms: ``potential_energy``, ``total_energy``, etc. (IMD-streamed in NAMD and GROMACS only)

Reviewer 2: Nitpick, but maybe say available as an option in NAMD and GROMACS only
```suggestion
* Energy terms: ``potential_energy``, ``total_energy``, etc. (available as an option in NAMD and GROMACS only)
```

  - Within-frame: :meth:`~MDAnalysis.core.groups.AtomGroup.center_of_mass`, :meth:`~MDAnalysis.core.groups.AtomGroup.radius_of_gyration`
  - Within-frame: :func:`~MDAnalysis.analysis.distances.distance_array` for pairwise distances between atom groups
  - Between-frames: :func:`~MDAnalysis.lib.distances.calc_bonds` for displacement calculations comparing consecutive frames
  - Between-frames: Frame-to-frame RMSD calculations using :func:`~MDAnalysis.analysis.rms.rmsd` with stored reference coordinates
  - Within-frame: Real-time monitoring using :meth:`~MDAnalysis.core.groups.AtomGroup.center_of_mass` for quality control checks

* **Accumulative analyses**: Building results incrementally across frames using :class:`~MDAnalysis.analysis.base.AnalysisBase` patterns, for example:

  - :class:`~MDAnalysis.analysis.rdf.InterRDF` - Frame-by-frame radial distribution function calculations
  - :class:`~MDAnalysis.analysis.dihedrals.Dihedral` - Dihedral angle accumulation for conformational analysis  
  - :class:`~MDAnalysis.analysis.lineardensity.LinearDensity` - Density profile building over streaming frames

> 

Reviewer 2: Thanks for writing this up, adds a ton of clarity for users

Read through the analysis classes, agree these are good examples of what will work.

One small suggestion, should specify that backend has to be serial
```suggestion
.. note::
    Passing any backend other than ``backend="serial"`` to these analysis classes will cause them to fail since streams are single-pass and forward-only.
```
